api/src/main/java/io/minio/MinioClient.java

```diff
@@ -2383,6 +2386,218 @@ public void copyObject(String bucketName, String objectName, Map<String,String>
+      }
+
+      if (size > MAX_PART_SIZE) {
+        long count = size / MAX_PART_SIZE;
+        long lastPartSize = size - (count * MAX_PART_SIZE);
+        if (lastPartSize > 0) {
+          count++;
+        } else {
+          lastPartSize = MAX_PART_SIZE;
+        }
+
+        if (lastPartSize < MIN_MULTIPART_SIZE && sources.size() != 1 && i != (sources.size() - 1)) {
+          throw new InvalidArgumentException("source " + src.bucketName() + "/" + src.objectName() + ": "
+            + "for multipart split upload of " + size
+            + ", last part size is less than " + MIN_MULTIPART_SIZE);
+        }
+
+        partsCount += (int)count;
+      } else {
+        partsCount++;
+      }
+
+      if (partsCount > MAX_MULTIPART_COUNT) {
+        throw new InvalidArgumentException("Compose sources create more than allowed multipart count "
+          + MAX_MULTIPART_COUNT);
+      }
+    }
+
+    if (partsCount == 1) {
+      ComposeSource src = sources.get(0);
+      copyObject(bucketName, objectName, headerMap, sse, src.bucketName(), src.objectName(), src.sse(),
+          src.copyConditions());
+      return;
+    }
+    if (headerMap == null) {
+      headerMap = new HashMap<>();
+    }
+
+    if (sse != null) {
+      headerMap = sse.headers();
+    }
+
+    String uploadId = initMultipartUpload(bucketName, objectName, headerMap);
+
+    int partNumber = 0;
+    Part[] totalParts = new Part[partsCount];
+    try {
+      for (int i = 0; i < sources.size(); i++) {
+        ComposeSource src = sources.get(i);
+
+        long size = src.objectSize();
+        if (src.length() != null) {
+          size = src.length();
+        }
+        if (src.offset() != null) {
+          size -= src.offset();
+        }
+
+        long offset = 0;
+        if (src.offset() != null) {
+          offset = src.offset();
+        }
+
+        if (size <= MAX_PART_SIZE) {
+          partNumber++;
+
+          Map<String, String> headers = src.headers();
+          if (src.length() != null) {
+            headers.put("x-amz-copy-source-range", "bytes=" + offset + "-" + (offset + src.length()));
+          } else if (src.offset() != null) {
+            headers.put("x-amz-copy-source-range", "bytes=" + offset + "-" + (offset + size - 1));
+          }
+          headerMap.putAll(headers);
+          String eTag = uploadPartCopy(bucketName, objectName, uploadId, partNumber, headerMap);
+
+          totalParts[partNumber - 1] = new Part(partNumber, eTag);
+          continue;
+        }
+
+        while (size > 0) {
+          partNumber++;
+
+          long startBytes = offset;
+          long endBytes = startBytes + MAX_PART_SIZE;
+          if (size < MAX_PART_SIZE) {
+            endBytes = startBytes + size;
+          }
+
+          Map<String, String> headers = src.headers();
+          headers.put("x-amz-copy-source-range", "bytes=" + startBytes + "-" + endBytes);
```

Comment:
You need to add SSE headers here.

Reply:
added

Reply:
Nope. You haven't added SSE headers here.